File: yeti/messenger.py
import json
import threading
import socket
import selectors
import traceback


class Messenger:

    running = False

    udp_port = None
    tcp_port = None
    address = None

    def __init__(self, messenger_id):
        self.message_types = {}
        self.address_book = {}

        self.tcp_sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self.udp_sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)

        self.messenger_id = messenger_id

        # Init message receive threads
        self.parent_thread = threading.current_thread()
        self.tcp_thread = threading.Thread(target=self._tcp_receive_loop, name="messenger_{}_tcp".format(messenger_id))
        self.udp_thread = threading.Thread(target=self._udp_receive_loop, name="messenger_{}_udp".format(messenger_id))

        # Init builtin message type
        self.register_message_type("address_resolution_request", self._on_address_resolution_request)

    def start_server(self, address, udp_port, tcp_port):
        self.address = address
        self.udp_port = udp_port
        self.tcp_port = tcp_port

        self.register_messenger_address(self.messenger_id, address, udp_port, tcp_port)
        self.running = True
        self.tcp_thread.start()
        self.udp_thread.start()
        print("Started {} on {}, {}, {}".format(self.messenger_id, address, udp_port, tcp_port))

    def _check_alive(self):
        self.running = self.running and self.parent_thread.is_alive()

    def stop_server(self):
        self.running = False
        if self.udp_thread.is_alive():
            self.udp_thread.join()
        if self.tcp_thread.is_alive():
            self.tcp_thread.join()

    def _tcp_receive_loop(self):
        self.tcp_sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        self.tcp_sock.bind(("0.0.0.0", self.tcp_port))
        self.tcp_sock.setblocking(False)
        self.tcp_sock.listen(10)
        selector = selectors.DefaultSelector()
        selector.register(self.tcp_sock, selectors.EVENT_READ)
        while self.running:
            self._check_alive()
            events = selector.select(1)
            for _, _ in events:
                try:
                    conn, addr = self.tcp_sock.accept()
                    conn.setblocking(True)
                    conn.settimeout(0.005)
                    str_dat = conn.recv(1024)

                    message = self._parse_message(str_dat.decode("utf-8"))
                    ret = message["handler"](message["data"])
                    if ret is not None:
                        conn.send(json.dumps(ret).encode("utf-8"))
                finally:
                    conn.close()
        print("tcp loop terminated")
        self.tcp_sock.shutdown(socket.SHUT_RDWR)
        self.tcp_sock.close()

    def _udp_receive_loop(self):
        self.udp_sock.bind(("0.0.0.0", self.udp_port))
        self.udp_sock.setblocking(False)
        selector = selectors.DefaultSelector()
        selector.register(self.udp_sock, selectors.EVENT_READ)
        while self.running:
            self._check_alive()
            events = selector.select(1)
            for _, _ in events:
                try:
                    str_dat = self.udp_sock.recv(1024)
                    message = self._parse_message(str_dat.decode("utf-8"))
                    message["handler"](message["data"])
                except Exception as e:
                    print(traceback.format_exc())
        print("udp loop terminated")
        self.udp_sock.close()

    def _parse_message(self, msg):
        if msg == "":
            return None
        data = json.loads(msg)
        for k in ["msg_type"]:
            if k not in data:
                raise ValueError("Invalid message, missing {} field.".format(k))
        if data["msg_type"] not in self.message_types:
            raise ValueError("Invalid message type: was {}, should be one of {}".format(data["msg_type"], [k for k in self.message_types.keys()]))
        message_type = data["msg_type"]
        data["handler"] = self.message_types[message_type]
        return data

    def register_messenger_address(self, messenger_id, address=None, udp_port=None, tcp_port=None):
        if address is None:
            address = self.address
            udp_port = self.udp_port
            tcp_port = self.tcp_port
        self.address_book[messenger_id] = (address, udp_port, tcp_port)

    def register_message_type(self, message_type, callback):
        self.message_types[message_type] = callback

    def send_message(self, message_type, data, client_id, blocking=False):
        address, udp_port, tcp_port = self.resolve_address(client_id)
        message_dict = {
            "msg_type": message_type,
            "data": data,
        }
        data = json.dumps(message_dict)
        if blocking:
            sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            sock.setblocking(True)
            sock.settimeout(3.0)
            sock.connect((address, tcp_port))
            sock.send(data.encode("utf-8"))
            ret_data = sock.recv(1024)
            sock.shutdown(socket.SHUT_RDWR)
            sock.close()
            s = ret_data.decode("utf-8")
            if s == "":
                return None
            return json.loads(s)
        else:
            self.udp_sock.sendto(data.encode("utf-8"), (address, udp_port))

    def _on_address_resolution_request(self, data):
        response = {"request_blacklist": data["request_blacklist"]}
        try:
            response["addr"], \
            response["udp_port"], \
            response["tcp_port"] = self.resolve_address(data["messenger_id"], data["request_blacklist"])
        except CouldNotResolveMessengerException:
            pass
        return response

    def resolve_address(self, messenger_id, request_blacklist=None):
        """
        Resolves the address and port number of a messenger.
        First tries to use cache, then scans all other registered messengers.
        """
        if messenger_id not in self.address_book:
            # Setup a request blacklist so we don't recurse over the network.
            if request_blacklist is None:
                request_blacklist = []
            for client_id in self.address_book:
                if (self.address, self.udp_port, self.tcp_port) == self.address_book[client_id]:
                    request_blacklist.append(client_id)
            for client_id in self.address_book:
                if client_id in request_blacklist:
                    continue
                try:
                    response = self.send_message("address_resolution_request",
                                                 {"messenger_id": messenger_id,
                                                  "request_blacklist": request_blacklist},
                                                 client_id, blocking=True)
                    if "addr" in response:
                        break
                    else:
                        request_blacklist.clear()
                        request_blacklist.extend(response["request_blacklist"])
                except (ConnectionError, socket.timeout) as e:
                    request_blacklist.append(client_id)
            else:
                raise CouldNotResolveMessengerException(messenger_id)
            self.address_book[messenger_id] = response["addr"], response["udp_port"], response["tcp_port"]
        return self.address_book[messenger_id]


class CouldNotResolveMessengerException(Exception):

    def __init__(self, messenger_id):
        super().__init__("{}".format(messenger_id))
        self.monitor_id = messenger_id

File: java/src/yeti/Messenger.java
package yeti;

import java.io.BufferedReader;
import java.io.DataOutputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.InetAddress;
import java.net.ServerSocket;
import java.net.Socket;
import java.net.SocketException;
import java.nio.ByteBuffer;
import java.nio.channels.DatagramChannel;
import java.nio.channels.Selector;
import java.nio.channels.ServerSocketChannel;
import java.nio.channels.SocketChannel;
import java.nio.charset.Charset;
import java.net.InetSocketAddress;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.TreeMap;
import java.util.function.Function;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

public class Messenger {
	
	private class Address{
		public String addr;
		public int udpPort;
		public int tcpPort;
		
		Address(String addr, int udp_port, int tcp_port){
			this.addr = addr;
			this.udpPort = udp_port;
			this.tcpPort = tcp_port;
		}
	}
	
	boolean running = false;
	
	int udpPort;
	int tcpPort;
	String address;
	
	String messengerId;
	
	TreeMap<String, Address> addressBook;
	HashMap<String, Function<JSONObject, JSONObject>> callbacks;
	
	DatagramSocket udpSock;
	
	Thread parentThread;
	Thread udpThread;
	Thread tcpThread;
	
	Messenger(String messengerId) throws IOException{
		this.messengerId = messengerId;
		addressBook = new TreeMap<String, Address>();
		callbacks  = new HashMap<String, Function<JSONObject, JSONObject>>();
		
		udpSock = new DatagramSocket();
		
		udpThread = new Thread(){public void run(){udpReceiveLoop();}};
		tcpThread = new Thread(){public void run(){tcpReceiveLoop();}};
		
		registerMessageType("address_resolution_request", this::onAddressResolutionRequest);
		
	}
	
	public void startServer(String addr, int udpPort, int tcpPort){
		parentThread = Thread.currentThread();
		
		this.udpPort=udpPort;
		this.tcpPort=tcpPort;
		this.address = addr;
		registerMessengerAddress(this.messengerId, addr, udpPort, tcpPort);
		
		running = true;
		udpThread.start();
		tcpThread.start();
		System.out.println("Started " + this.messengerId + " on " + addr + ", " + udpPort + ", " + tcpPort);
	}
	
	public void stopServer() throws InterruptedException{
		System.out.println("Stopping " + this.messengerId);
		running = false;
		
		if (udpThread.isAlive())
			udpThread.join();
		if (tcpThread.isAlive())
			tcpThread.join();
		System.out.println("Stopped " + this.messengerId);
	}
	
	private void udpReceiveLoop(){
		System.out.println("Starting udp loop for " + this.messengerId);
		
		try {
			DatagramChannel channel = DatagramChannel.open();
			channel.socket().bind(new InetSocketAddress(udpPort));
			channel.configureBlocking(false);
			ByteBuffer packet = ByteBuffer.allocate(1024);
			packet.clear();
			while (running && parentThread.isAlive()){
				if(channel.receive(packet) != null){
					String json_text = new String(packet.array(), "ASCII");
					JSONObject message = new JSONObject(json_text);
					String msgType = (String) message.get("msg_type");
					callbacks.get(msgType).apply(message.getJSONObject("data"));
					packet.clear();
				}
				Thread.sleep(10);
			}
			channel.close();
			System.out.println("Stopped udp loop for " + this.messengerId);
		} catch (IOException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		} catch (InterruptedException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
	}
	
	private void tcpReceiveLoop(){
		System.out.println("Starting tcp loop for " + this.messengerId);
		try {
			ServerSocketChannel channel = ServerSocketChannel.open();
			channel.socket().bind(new InetSocketAddress(tcpPort));
			channel.configureBlocking(false);
			
			ByteBuffer packet = ByteBuffer.allocate(1024);
			packet.clear();
			
			while (running && parentThread.isAlive()){
				SocketChannel socketChannel = channel.accept();
				
				if (socketChannel != null){
					socketChannel.configureBlocking(false);
					int msg_len = 0;
					int msg_len_len = 0;
					int bytes_read = 0;
					while (true){
						int res = socketChannel.read(packet);
						bytes_read += res;
						if (res == 0)
							Thread.sleep(10);
						else if (res <= 0)
							throw new IOException(); 
						String buf = new String(packet.array(), "utf-8");
						int ind = buf.indexOf("!");
						if (ind > 0){
							msg_len_len = ind;
							msg_len = Integer.parseInt(buf.substring(1, ind));
							break;
						}
					}
					while (bytes_read < msg_len + msg_len_len){
						int res = socketChannel.read(packet);
						if (res == 0)
							Thread.sleep(10);
						bytes_read += res;
					}
					String json_text = (new String(packet.array(), "utf-8")).substring(msg_len_len+1);
	
					JSONObject message = new JSONObject(json_text);
					String msgType = (String) message.get("msg_type");
					JSONObject response = callbacks.get(msgType).apply(message.getJSONObject("data"));
					byte[] responseData = response.toString().getBytes("utf-8");
					
					packet.clear();
					packet.put(("|" + responseData.length + "!").getBytes("utf-8"));
					packet.put(responseData);
					packet.flip();
					
					while (packet.hasRemaining()){
						socketChannel.write(packet);
					}
					packet.clear();
					socketChannel.close();
				}
				Thread.sleep(10);
			}
			channel.close();
			System.out.println("Stopping tcp loop for " + this.messengerId);
		} catch (IOException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		} catch (InterruptedException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
	}
	
	public void registerMessageType(String messageType, Function<JSONObject, JSONObject> callback){
		callbacks.put(messageType, callback);
	}
	
	public void registerMessengerAddress(String messengerId, String addr, int udpPort, int tcpPort){
		addressBook.put(messengerId, new Address(addr, udpPort, tcpPort));
	}
	
	public JSONObject sendMessage(String messageType, JSONObject data, String clientId, boolean blocking) throws IOException, CouldNotResolveMessengerException{
		Address address = resolveAddress(clientId);
		JSONObject message = new JSONObject();
		message.put("msg_type", messageType);
		message.put("data", data);
		byte messageData[] = message.toString().getBytes("utf-8");
		
		if (blocking){
			Socket sock = new Socket();
			sock.connect(new InetSocketAddress(address.addr, address.tcpPort));
			DataOutputStream outToServer = new DataOutputStream(sock.getOutputStream());
			BufferedReader inFromServer = new BufferedReader(new InputStreamReader(sock.getInputStream(), "UTF-8"));
			
			outToServer.writeBytes("|" + messageData.length + "!");
			outToServer.write(messageData);
			outToServer.flush();
			String buffer = "";
			int msgLen = 0;
			while (true) {
				char c = (char) inFromServer.read();
				if (c == '|')
					continue;
				if (c == '!'){
					msgLen = Integer.parseInt(buffer);
					break;
				}
				buffer += c;
			}
			String jsonResponseStr = "";
			while (jsonResponseStr.length() < msgLen)
				jsonResponseStr += (char) inFromServer.read();
			sock.close();
			return new JSONObject(jsonResponseStr);
		}
		else {
			DatagramPacket packet = new DatagramPacket(messageData, messageData.length, InetAddress.getByName(address.addr), address.udpPort);
			udpSock.send(packet);
			return new JSONObject();
		}
	}
	
	private JSONObject onAddressResolutionRequest(JSONObject message){
		// Repack responseBlacklist from a JSONArray to an ArrayList
		ArrayList<String> requestBlacklist = new ArrayList<String>();
		JSONArray jsonRequestBlacklist = message.getJSONArray("request_blacklist");
		for (int i = 0; i < jsonRequestBlacklist.length(); i++) {
			requestBlacklist.add(jsonRequestBlacklist.getString(i));
		}
		
		// Resolve the address.
		Address result;
		try {
			result = resolveAddress(message.getString("messenger_id"), requestBlacklist);
		} catch (JSONException | CouldNotResolveMessengerException e) {
			result = null;
		}
		
		// Pack the response
		JSONObject responseData = new JSONObject();
		if (result != null){
			responseData.put("addr", result.addr);
			responseData.put("udp_port", result.udpPort);
			responseData.put("tcp_port", result.tcpPort);
		}
		responseData.put("request_blacklist", requestBlacklist);
		return responseData;
	}
	
	private Address resolveAddress(String messengerId) throws CouldNotResolveMessengerException{
		return resolveAddress(messengerId, new ArrayList<String>());
	}
	
	private Address resolveAddress(String messengerId, ArrayList<String> requestBlacklist) throws 
	CouldNotResolveMessengerException {
		
		if (!addressBook.containsKey(messengerId)){
			// Blacklist any client ids that point to us.
			for (String clientId : addressBook.keySet()){
				Address address = addressBook.get(clientId);
				if (address.addr.equals(this.address) && address.udpPort == this.udpPort && address.tcpPort == this.tcpPort)
					requestBlacklist.add(clientId);
			}
			// Search over all known clients for a valid address
			for (String clientId : addressBook.keySet()){
				if (requestBlacklist.contains(clientId))
					continue;
				try{
					JSONObject requestData = new JSONObject();
					requestData.put("messenger_id", messengerId);
					requestData.put("request_blacklist", requestBlacklist);
					JSONObject responseData = sendMessage("address_resolution_request", requestData, clientId, true);
					if (responseData.has("addr")){
						Address newAddress = new Address(responseData.getString("addr"),
								responseData.getInt("udp_port"),
								responseData.getInt("tcp_port"));
						addressBook.put(messengerId, newAddress);
						return newAddress;
					}
					else {
						// Reset requestBlacklist to new list from client
						requestBlacklist.clear();
						JSONArray newRequestBlacklist = responseData.getJSONArray("request_blacklist");
						for (int i = 0; i < newRequestBlacklist.length(); i++){
							requestBlacklist.add(newRequestBlacklist.getString(i));
						}
					}
				} catch (IOException e){
					requestBlacklist.add(clientId);
				}
			}	
			throw new Messenger.CouldNotResolveMessengerException(messengerId);
		}
		return addressBook.get(messengerId);
	}
	
	public class CouldNotResolveMessengerException extends Exception{
		private String messengerId;
		
		public CouldNotResolveMessengerException(String messengerId){
			super(messengerId);
			this.messengerId = messengerId;
		}
	}
}